### PR TITLE
core:crypto: Add X25519 support

### PR DESCRIPTION
**Pull request by sohhass**
Hey folks!

This PR adds X25519 support for OP-TEE as defined in GlobalPlatform TEE Internal Core API v1.2 using libtomcrypt as the backend crypto implementation.

Since the Curve25519 implementation does not support bignums, there is also a new key attribute type (octet reference) as defined in GlobalPlatform TEE Internal Core API v1.2 Table 6-15.

### CODE REVIEW

File: core/tee/tee_svc_cryp.c [hunk 1]
```diff
@@ -78,6 +79,11 @@ struct tee_cryp_obj_secret {
 #define ATTR_OPS_INDEX_BIGNUM     1
     /* Convert to/from value attribute depending on direction */
 #define ATTR_OPS_INDEX_VALUE      2
+    /* Convert to/from curve25519 attribute depending on direction */
+#define ATTR_OPS_INDEX_25519      3
```
Comment (jforissier) [inline core/tee/tee_svc_cryp.c:83]: Why is this needed? All other algorithms use bignum and value.

**Comment by jforissier** [comment]
For commit "libutee: add X25519 support": with one minor comment:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/tee/tee_svc_cryp.c (hunk 1 above)
Comment (sohhass) [reply core/tee/tee_svc_cryp.c:83]: > Why is this needed? All other algorithms use bignum and value.

The X(Ed)25519 curve operations arguments are  32-byte octet strings as defined in
[RFC 7748, pg 7](https://datatracker.ietf.org/doc/html/rfc7748#page-7). Also from TEE Internal Core API Specification v1.2, Table Table 6-15, `TEE_ATTR_X25519_*` is a binary octet string (while other asym crypto attrs are bignum). The underlying implementation of libtomcrypt for Curve25519 operations does not use bignums (AFAIK most of the library implementations follow this convention). The X25519 byte level encoding is `little-endian` according to the `RFC-7748`, and if I am not mistaken for bignums it is big-endian. If we go the bignum route, we should then add conversions from bignum to appropriate binary format before calling the `X25519` ops.

Comment (jforissier) [reply core/tee/tee_svc_cryp.c:83]: Makes sense, thanks for the details. Let's keep this new define.

File: core/tee/tee_svc_cryp.c [hunk 2]
```diff
@@ -871,6 +898,121 @@ static void op_attr_value_clear(void *attr)
...
+	if (s < key_size || !buffer)
+		return TEE_ERROR_SHORT_BUFFER;
+
+	return copy_to_user(buffer, *key, key_size);
+}
+
+static TEE_Result op_attr_25519_to_binary(void *attr, void *data,
+					  size_t data_len, size_t *offs)
+{
+	TEE_Result res;
```
Comment (etienne-lms) [inline core/tee/tee_svc_cryp.c:942]:
```c
	TEE_Result res = TEE_ERROR_GENERIC;
	size_t next_offs = 0;

File: core/tee/tee_svc_cryp.c [hunk 3]
```diff
@@ -871,6 +898,121 @@ static void op_attr_value_clear(void *attr)
...
+	(*offs) = next_offs;
+
+	return TEE_SUCCESS;
+}
+
+static bool op_attr_25519_from_binary(void *attr, const void *data,
+				      size_t data_len, size_t *offs)
+{
+	uint8_t **key = attr;
+	uint32_t s;
```
Comment (etienne-lms) [inline core/tee/tee_svc_cryp.c:965]: = 0;

File: core/tee/tee_svc_cryp.c [hunk 4]
```diff
@@ -871,6 +898,121 @@ static void op_attr_value_clear(void *attr)
...
+}
+
+static void op_attr_25519_clear(void *attr)
+{
+	uint8_t **key = attr;
+	size_t _size = KEY_SIZE_BYTES_25519;
+
+	assert(*key);
+
+	memzero_explicit(*key, _size);
```
Comment (etienne-lms) [inline core/tee/tee_svc_cryp.c:1013]: `memzero_explicit(*key, KEY_SIZE_BYTES_25519);` is fine (IMHO)

Comment (etienne-lms) [inline core/tee/tee_svc_cryp.c:1013]: `memzero_explicit(*key, KEY_SIZE_BYTES_25519);` is fine (IMHO)

File: core/tee/tee_svc_cryp.c [hunk 5]
```diff
@@ -3226,6 +3420,41 @@ TEE_Result syscall_cryp_derive_key(unsigned long state,
 		crypto_acipher_free_ecc_public_key(&peer_key);
 		crypto_acipher_free_ecc_public_key(&peer_eph_key);
 	}
+#endif
+#if defined(CFG_CRYPTO_X25519)
+	else if (cs->algo == TEE_ALG_X25519) {
+		uint8_t *x25519_pub_key;
+		uint8_t *pt_secret;
+		unsigned long pt_secret_len;
```
Comment (etienne-lms) [inline core/tee/tee_svc_cryp.c:3428]:
```c
		uint8_t *x25519_pub_key = NULL;
		uint8_t *pt_secret = NULL;
		unsigned long pt_secret_len = 0;

Comment (etienne-lms) [inline core/tee/tee_svc_cryp.c:3428]:
```c
		uint8_t *x25519_pub_key = NULL;
		uint8_t *pt_secret = NULL;
		unsigned long pt_secret_len = 0;

File: core/lib/libtomcrypt/src/sub.mk [hunk 6]
```diff
@@ -6,3 +6,7 @@ subdirs-$(_CFG_CORE_LTC_ACIPHER) += math
 subdirs-y += misc
 subdirs-y += modes
 subdirs-$(_CFG_CORE_LTC_ACIPHER) += pk
+
+ifeq ($(_CFG_CORE_LTC_ACIPHER),n)
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/src/sub.mk:10]: prefer: `ifneq ($(_CFG_CORE_LTC_ACIPHER),y)`

File: core/lib/libtomcrypt/x25519.c [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+	memcpy(ltc_public_key.pub, public_key,
+		    X25519_KEY_SIZE_BYTES);
+
+	memcpy(ltc_private_key.priv, private_key->priv,
+		    X25519_KEY_SIZE_BYTES);
+
+	if (x25519_shared_secret(&ltc_private_key, &ltc_public_key,
+		    secret, secret_len) != CRYPT_OK)
+				return TEE_ERROR_BAD_PARAMETERS;
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:104]:
```c
	if (x25519_shared_secret(&ltc_private_key, &ltc_public_key,
				 secret, secret_len) != CRYPT_OK)
		return TEE_ERROR_BAD_PARAMETERS;

File: core/lib/libtomcrypt/x25519.c [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+	if (sizeof(ltc_public_key.pub) < X25519_KEY_SIZE_BYTES
+		|| sizeof(ltc_private_key.priv) < X25519_KEY_SIZE_BYTES)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+
+	memcpy(ltc_public_key.pub, public_key,
+		    X25519_KEY_SIZE_BYTES);
+
+	memcpy(ltc_private_key.priv, private_key->priv,
+		    X25519_KEY_SIZE_BYTES);
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:100]: each of the 2 fits in a single line
```c
	memcpy(ltc_public_key.pub, public_key, X25519_KEY_SIZE_BYTES);
	memcpy(ltc_private_key.priv, private_key->priv, X25519_KEY_SIZE_BYTES);

File: core/lib/libtomcrypt/x25519.c [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+	if (x25519_shared_secret(&ltc_private_key, &ltc_public_key,
+		    secret, secret_len) != CRYPT_OK)
+				return TEE_ERROR_BAD_PARAMETERS;
+
+	memset(&ltc_private_key, 0, sizeof(ltc_private_key));
+	memset(&ltc_public_key, 0, sizeof(ltc_public_key));
+	return TEE_SUCCESS;
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:108]: add an empty line above `return` instruction.

File: core/lib/libtomcrypt/x25519.c [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+	};
+	curve25519_key ltc_public_key = {
+		.type = PK_PUBLIC,
+		.algo = PKA_X25519
+	};
+
+	if (private_key == NULL
+		|| public_key == NULL
+		|| secret == NULL
+		|| secret_len == NULL)
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:87]:
```c
	if (!private_key || !public_key || !secret || !secret_len)
		return TEE_ERROR_BAD_PARAMETERS;

	COMPILE_TIME_ASSERT(sizeof(ltc_public_key.pub) ==
			    X25519_KEY_SIZE_BYTES &&
			    sizeof(ltc_public_key.priv) ==
			    X25519_KEY_SIZE_BYTES);
```

File: core/lib/libtomcrypt/x25519.c [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+TEE_Result crypto_acipher_x25519_shared_secret(struct x25519_keypair
+					       *private_key,
+					       void *public_key,
+					       void *secret,
+					       unsigned long *secret_len)
+{
+
+	curve25519_key ltc_private_key = {
+		.type = PK_PRIVATE,
+		.algo = PKA_X25519
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:77]: prefer with a terminal coma for each line (ditto line 81)

File: core/lib/libtomcrypt/x25519.c [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+	return TEE_SUCCESS;
+}
+
+TEE_Result crypto_acipher_x25519_shared_secret(struct x25519_keypair
+					       *private_key,
+					       void *public_key,
+					       void *secret,
+					       unsigned long *secret_len)
+{
+
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:74]: remove empty line

File: core/lib/libtomcrypt/x25519.c [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+	memset(&ltc_tmp_key, 0, sizeof(ltc_tmp_key));
+
+	if (x25519_make_key(&dummy, find_prng("prng_crypto"),
+	    &ltc_tmp_key)!= CRYPT_OK)
+			return TEE_ERROR_BAD_PARAMETERS;
+
+	if (key_size_bytes < sizeof(ltc_tmp_key.pub)
+		|| key_size_bytes < sizeof(ltc_tmp_key.priv))
+		return TEE_ERROR_BAD_PARAMETERS;
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:59]: if (key_size_bytes < sizeof(ltc_tmp_key.pub) ||
	    key_size_bytes < sizeof(ltc_tmp_key.priv))
		return TEE_ERROR_BAD_PARAMETERS;

File: core/lib/libtomcrypt/x25519.c [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+			return TEE_ERROR_OUT_OF_MEMORY;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result crypto_acipher_gen_x25519_key(struct x25519_keypair *key,
+					 size_t key_size)
+{
+	curve25519_key ltc_tmp_key = { };
+	size_t key_size_bytes = (size_t)(key_size / 8);
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:45]: remove cast

File: core/lib/libtomcrypt/x25519.c [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+	prng_state dummy;
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:46]: `= { };`

File: core/lib/libtomcrypt/x25519.c [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+	if (key_size_bytes != X25519_KEY_SIZE_BYTES)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	memset(&ltc_tmp_key, 0, sizeof(ltc_tmp_key));
+
+	if (x25519_make_key(&dummy, find_prng("prng_crypto"),
+	    &ltc_tmp_key)!= CRYPT_OK)
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:54]:
```c
	if (x25519_make_key(&dummy, find_prng("prng_crypto"), &ltc_tmp_key) !=
	    CRYPT_OK)

File: core/lib/libtomcrypt/x25519.c [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+			return TEE_ERROR_BAD_PARAMETERS;
+
+	if (key_size_bytes < sizeof(ltc_tmp_key.pub)
+		|| key_size_bytes < sizeof(ltc_tmp_key.priv))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	memcpy(key->pub, ltc_tmp_key.pub, sizeof(ltc_tmp_key.pub));
+	memcpy(key->priv, ltc_tmp_key.priv, sizeof(ltc_tmp_key.priv));
+	memset(&ltc_tmp_key, 0, sizeof(ltc_tmp_key));
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:63]: candidate for `memzero_explicit()`?

File: core/lib/libtomcrypt/x25519.c [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+/* X25519 key is an octet string of 32 bytes */
+#define X25519_KEY_SIZE_BYTES UL(32)
+
+TEE_Result crypto_acipher_alloc_x25519_keypair(struct x25519_keypair *key,
+					       size_t key_size)
+{
+
+	size_t key_size_bytes = (size_t)(key_size / 8);
+
+	if (key == NULL)
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:24]: prefer `if (!key)` for consistency in optee_os

File: core/lib/libtomcrypt/x25519.c [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+
+	memset(key, 0, sizeof(*key));
+
+	if (key_size_bytes != X25519_KEY_SIZE_BYTES)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	key->priv = calloc(1, key_size_bytes);
+	key->pub = calloc(1, key_size_bytes);
+
+	if (key->priv == NULL || key->pub == NULL)
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:35]:
```c
	if (!key->priv || !key->pub) {
		free(key->priv);
		free(key->pub);
		return TEE_ERROR_OUT_OF_MEMORY;
	}

File: core/lib/libtomcrypt/x25519.c [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+#include "acipher_helpers.h"
+
+/* X25519 key is an octet string of 32 bytes */
+#define X25519_KEY_SIZE_BYTES UL(32)
+
+TEE_Result crypto_acipher_alloc_x25519_keypair(struct x25519_keypair *key,
+					       size_t key_size)
+{
+
+	size_t key_size_bytes = (size_t)(key_size / 8);
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/x25519.c:22]: cast not needed

File: core/tee/tee_svc_cryp.c (hunk 2 above)
Comment (sohhass) [reply core/tee/tee_svc_cryp.c:942]: `op_attr_25519_to_binary` should we initialize all `*_to_binary` (bignum, value) functions? was here for just consistency

File: core/tee/tee_svc_cryp.c (hunk 3 above)
Comment (sohhass) [reply core/tee/tee_svc_cryp.c:965]: `op_attr_25519_from_binary`:  ditto as `op_attr_25519_to_binary` comment

File: core/lib/libtomcrypt/x25519.c (hunk 19 above)
Comment (sohhass) [reply core/lib/libtomcrypt/x25519.c:35]: good catch, will fix!

File: core/tee/tee_svc_cryp.c [hunk 21]
```diff
@@ -871,6 +898,121 @@ static void op_attr_value_clear(void *attr)
...
+	uint64_t key_size = (uint64_t)KEY_SIZE_BYTES_25519;
+
+	res = copy_from_user(&s, size, sizeof(s));
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = copy_to_user(size, &key_size, sizeof(key_size));
+	if (res != TEE_SUCCESS)
+		return res;
+	if (!key_size)
```
Comment (etienne-lms) [inline core/tee/tee_svc_cryp.c:930]: is it possible that `key_size == 0` here? Did you mean `if (!s)`?

File: core/tee/tee_svc_cryp.c (hunk 2 above)
Comment (etienne-lms) [reply core/tee/tee_svc_cryp.c:942]: We should fix other `*_to_binary()` functions, but that may be the subject for another P-R. As here you introduce new code, prefer have it following optee_os coding style from the start.

File: core/lib/libtomcrypt/x25519.c (hunk 15 above)
Comment (sohhass) [reply core/lib/libtomcrypt/x25519.c:46]: @etienne-lms Thanks for the review! One comment on this one, the struct `prng_state` here is passed since the libtomcrypt X25519 implementation expects a valid object due to this check `LTC_ARGCHK(prng != NULL)` (https://github.com/OP-TEE/optee_os/blob/master/core/lib/libtomcrypt/src/pk/x25519/x25519_make_key.c#L30), in reality it is never used as OP-TEE has its own `prng_crypto_desc` ("prng_crypto") and all the `prng`functions declare `prng_state` as `__unused`. 

Also I am not sure if it is safe to initialize `prng_state` this way since the initialization needs to properly handle the mutex. 

Making this change produces this error during key generation:
```
D/LD:  ldelf:168 ELF (cb3e5ba0-adf1-11e0-998b-0002a5d5c51b) at 0x40014000
F/TC:? 0 trace_syscall:151 syscall #8 (syscall_check_access_rights)
F/TC:? 0 trace_syscall:151 syscall #27 (syscall_cryp_obj_alloc)
F/TC:? 0 trace_syscall:151 syscall #54 (syscall_obj_generate_key)
E/TC:1   Dead canary at end of 'stack_abt[3]' (0xe1a01fc)
E/TC:1   Panic at core/kernel/thread.c:124 <thread_check_canaries>
E/TC:1   TEE load address @ 0xe100000
E/TC:1   Call stack:
E/TC:1    0x0e108900
E/TC:1    0x0e113ec8
E/TC:1    0x0e116e58
E/TC:1    0x0e106a5c
```

One option is to remove this check `LTC_ARGCHK(prng != NULL)` from libtomcrypt implementation, but I avoid making any changes to the library.
Also note, this check is not present in other asymmetric crypto key generation functions, so this case is only valid for `X25519`.

File: core/tee/tee_svc_cryp.c (hunk 21 above)
Comment (sohhass) [reply core/tee/tee_svc_cryp.c:930]: thanks for pointing out , this check `if (!key_size)` is not needed, will remove!

File: core/lib/libtomcrypt/x25519.c (hunk 15 above)
Comment (jenswi-linaro) [reply core/lib/libtomcrypt/x25519.c:46]: I believe removing the `LTC_ARGCHK(prng != NULL)` check is the best option. I don't know how large `prng_state` is, but I'm afraid it can become quite large in certain configurations.

Comment (sohhass) [reply core/lib/libtomcrypt/x25519.c]: Thanks for the comment, `LTC_ARGCHK(prng != NULL)` removed

**Comment by sohhass** [comment]
@etienne-lms thanks for the your review, all your comments are addressed in the latest changes and rebased on top of #5301

**Comment by jforissier** [comment]
> rebased on top of #5301

We still need to figure how to sort these conflicting IDs, 5301 is not the right way it seems.

**Comment by sohhass** [comment]
> > rebased on top of #5301
> 
> We still need to figure out how to address the issue of conflicting IDs, 5301 is not the right way it seems.

Currently OP-TEE does not defineAPI Version headers that would allow compile time selection of specific version. This is defined in TEE Internal Core API v1.2 section 3.1.1

File: lib/libutee/include/tee_api_defines.h [hunk 22]
```diff
@@ -289,7 +294,7 @@
 #define TEE_ECC_CURVE_NIST_P384             0x00000004
 #define TEE_ECC_CURVE_NIST_P521             0x00000005
 #define TEE_ECC_CURVE_SM2                   0x00000300
-
+#define TEE_ECC_CURVE_25519                 0x00000300
```
[the remark this answers is not in the document]
Comment (jforissier) [reply lib/libutee/include/tee_api_defines.h:297]: Move up one line to keep numerical order

File: core/crypto.mk [hunk 23]
```diff
@@ -226,6 +228,7 @@ _CFG_CORE_LTC_HASH := $(call ltc-one-enabled, MD5 SHA1 SHA224 SHA256 SHA384 \
 _CFG_CORE_LTC_MAC := $(call ltc-one-enabled, HMAC CMAC CBC_MAC)
 _CFG_CORE_LTC_CBC := $(call ltc-one-enabled, CBC CBC_MAC)
 _CFG_CORE_LTC_ASN1 := $(call ltc-one-enabled, RSA DSA ECC)
+_CFG_CORE_LTC_EC25519 := $(call ltc-one-enabled, X25519 ED25519)
```
Comment (jforissier) [inline core/crypto.mk:231]: Do we really want `ED25519` here? It doesn't seem that `CFG_CRYPTO_ED25519` (and therefore `_CFG_CORE_LTC_ED25519`) is supported in this PR?

File: core/lib/libtomcrypt/acipher_helpers.h [hunk 24]
```diff
@@ -35,7 +35,7 @@ static inline TEE_Result convert_ltc_verify_status(int ltc_res, int ltc_stat)
 	}
 }
 
-#ifdef CFG_CRYPTOLIB_NAME_tomcrypt
+#ifdef _CFG_CORE_LTC_ECC
```
Comment (jforissier) [inline core/lib/libtomcrypt/acipher_helpers.h:38]: This is not strictly related to 25519 support, please make a separate patch for this fix.

File: core/lib/libtomcrypt/src/sub.mk [hunk 25]
```diff
@@ -6,3 +6,7 @@ subdirs-$(_CFG_CORE_LTC_ACIPHER) += math
 subdirs-y += misc
 subdirs-y += modes
 subdirs-$(_CFG_CORE_LTC_ACIPHER) += pk
+
+ifneq ($(_CFG_CORE_LTC_ACIPHER),y)
+subdirs-$(_CFG_CORE_LTC_EC25519) += pk
+endif
```
Comment (jforissier) [inline core/lib/libtomcrypt/src/sub.mk:12]: One single statement like this would be better IMO:
```make
ifneq (,$(filter y,$(_CFG_CORE_LTC_ACIPHER) $(_CFG_CORE_LTC_EC25519)))
subdirs-y += pk
endif
```

Slightly off-topic but perhaps we should consider allowing multiple occurrences of the same sub-directory? I mean the following change:
```diff
diff --git a/mk/subdir.mk b/mk/subdir.mk
index ffa61d4e..d280c309 100644
--- a/mk/subdir.mk
+++ b/mk/subdir.mk
@@ -21,6 +21,8 @@ srcs :=
 gen-srcs :=
 asm-defines-files :=
 
+uniq = $(if $1,$(firstword $1) $(call uniq,$(filter-out $(firstword $1),$1)))
+
 define process-subdir-srcs-y
 ifeq ($$(sub-dir),.)
 srcs                           += $1
@@ -174,8 +176,8 @@ this-out-dir :=
 asm-defines-y :=
 
 # Process subdirectories in current directory
-$$(foreach sd, $$(sub-subdirs), $$(eval $$(call process-subdir,$$(sd))))
+$$(foreach sd, $$(call uniq,$$(sub-subdirs)), $$(eval $$(call process-subdir,$$(sd))))
 endef #process-subdir
 
 # Top subdirectories
-$(foreach sd, $(subdirs), $(eval $(call process-subdir,$(sd))))
+$(foreach sd, $(call uniq,$(subdirs)), $(eval $(call process-subdir,$(sd))))
``` 
...would allow:
```make
subdirs-$(_CFG_CORE_LTC_ACIPHER) += pk
subdirs-$(_CFG_CORE_LTC_EC25519) += pk
```
Opinions welcome (I can create a separate PR)

File: core/lib/libtomcrypt/sub.mk [hunk 26]
```diff
@@ -100,6 +100,9 @@ endif
 ifneq (,$(filter y,$(_CFG_CORE_LTC_SM2_DSA) $(_CFG_CORE_LTC_SM2_PKE)))
    cppflags-lib-y += -DLTC_ECC_SM2
 endif
+ifeq ($(_CFG_CORE_LTC_X25519),y)
+   cppflags-lib-y += -DLTC_CURVE25519
+endif
```
Comment (jforissier) [inline core/lib/libtomcrypt/sub.mk:105]: Conditional not needed.
```
cppflags-lib-$(_CFG_CORE_LTC_X25519) += -DLTC_CURVE25519
```

File: core/crypto.mk (hunk 23 above)
Comment (sohhass) [reply core/crypto.mk:231]: not really, this was just a placeholder; `ED25519` will be a different PR, will remove from here

File: core/lib/libtomcrypt/src/sub.mk (hunk 25 above)
Comment (jenswi-linaro) [reply core/lib/libtomcrypt/src/sub.mk:12]: I like the idea @jforissier.

File: core/crypto.mk [hunk 27]
```diff
@@ -47,6 +47,7 @@ CFG_CRYPTO_ECC ?= y
 CFG_CRYPTO_SM2_PKE ?= y
 CFG_CRYPTO_SM2_DSA ?= y
 CFG_CRYPTO_SM2_KEP ?= y
+CFG_CRYPTO_X25519 ?= y
```
Comment (jforissier) [inline core/crypto.mk:50]: Add `ifeq ($(CFG_CRYPTOLIB_NAME),tomcrypt)` / `endif` and also `$(call force,CFG_CRYPTO_X25519,n,not supported by mbedtls)` in `core/core.mk` (same as for `CFG_CRYPTO_RSASSA_NA1` which is also supported by LTC only).

**Review by jforissier — COMMENTED** [review]
Some more comments

**Comment by sohhass** [comment]
> Some more comments

thanks, addressed!

File: core/lib/libtomcrypt/x25519.c [hunk 28]
```diff
@@ -0,0 +1,106 @@
...
+			    X25519_KEY_SIZE_BYTES);
+
+	memcpy(ltc_public_key.pub, public_key, X25519_KEY_SIZE_BYTES);
+	memcpy(ltc_private_key.priv, private_key->priv, X25519_KEY_SIZE_BYTES);
+
+	if (x25519_shared_secret(&ltc_private_key, &ltc_public_key,
+				 secret, secret_len) != CRYPT_OK)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	memzero_explicit(&ltc_private_key, sizeof(ltc_private_key));
```
Comment (jforissier) [inline core/lib/libtomcrypt/x25519.c:99]: I still don't understand this, `crypto_acipher_dh_shared_secret()` doesn't clear the private key for instance.

File: core/tee/tee_svc_cryp.c [hunk 29]
```diff
@@ -1,7 +1,8 @@
 // SPDX-License-Identifier: BSD-2-Clause
 /*
  * Copyright (c) 2014, STMicroelectronics International N.V.
- * Copyright (c) 2020, 2022 Linaro Limited
+ * Copyright (c) 2020, Linaro Limited
```
Comment (etienne-lms) [inline core/tee/tee_svc_cryp.c:4]: leave the Linaro copyright unchanged.

File: core/tee/tee_svc_cryp.c [hunk 30]
```diff
@@ -872,6 +899,116 @@ static void op_attr_value_clear(void *attr)
...
+		return res;
+
+	if (ADD_OVERFLOW(*offs, key_size, &next_offs))
+		return TEE_ERROR_OVERFLOW;
+
+	if (data && next_offs <= data_len)
+		memcpy((uint8_t *)data + *offs, *key, key_size);
+	*offs = next_offs;
+
+	return res;
```
Comment (etienne-lms) [inline core/tee/tee_svc_cryp.c:957]: return  TEE_SUCCESS;

**Comment by sohhass** [comment]
> The overall looks good to me. Some comments.

Thanks! comments addressed

File: core/lib/libtomcrypt/x25519.c (hunk 28 above)
Comment (jforissier) [reply core/lib/libtomcrypt/x25519.c:95]: @sohhass can you please explain why this is needed?

Comment (jforissier) [reply core/lib/libtomcrypt/x25519.c:95]: @sohhass can you please explain why this is needed?

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` with some minor comments.

File: core/tee/tee_svc_cryp.c [hunk 31]
```diff
@@ -872,6 +899,112 @@ static void op_attr_value_clear(void *attr)
...
+
+	assert(*key);
+
+	memzero_explicit(*key, KEY_SIZE_BYTES_25519);
+}
+
+static void op_attr_25519_free(void *attr)
+{
+	op_attr_25519_clear(attr);
+	free(*(uint8_t**)attr);
```
Comment (etienne-lms) [inline core/tee/tee_svc_cryp.c:1005]: checkpatch complains for a missing space char: `free(*(uint8_t **)attr);`

For consistency, maybe:
```c
	uint8_t **key = attr;

	op_attr_25519_clear(attr);
	free(*key);
```

File: core/tee/tee_svc_cryp.c [hunk 32]
```diff
@@ -1916,6 +2066,35 @@ static TEE_Result tee_svc_obj_generate_key_ecc(
...
+static TEE_Result
+tee_svc_obj_generate_key_x25519(struct tee_obj *o,
+				const struct tee_cryp_obj_type_props
+							*type_props,
+				uint32_t key_size,
+				const TEE_Attribute *params,
+				uint32_t param_count)
+{
+	TEE_Result res;
+	struct x25519_keypair *tee_x25519_key;
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/tee/tee_svc_cryp.c:2074]: to fix

**Review by etienne-lms — CHANGES_REQUESTED** [review]
There is still a coding style issue (reported by checkpatch)  and a missing local variable initialization. Once addressed, i'm fine with this P-R.

There is another issue CI complains on (when building optee_rust_examples_ext) but it is not related to this P-R (see https://github.com/OP-TEE/optee_os/issues/5361).

File: core/lib/libtomcrypt/src/pk/x25519/x25519_make_key.c [hunk 33]
```diff
@@ -27,7 +27,6 @@ int x25519_make_key(prng_state *prng, int wprng, curve25519_key *key)
 {
    int err;
 
-   LTC_ARGCHK(prng != NULL);
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/lib/libtomcrypt/src/pk/x25519/x25519_make_key.c:30]: @etienne-lms previous discussion: https://github.com/OP-TEE/optee_os/pull/5297#discussion_r864622745

[the remark this answers is not in the document]
Comment (jforissier) [reply core/lib/libtomcrypt/src/pk/x25519/x25519_make_key.c:30]: > @etienne-lms previous discussion: [#5297 (comment)](https://github.com/OP-TEE/optee_os/pull/5297#discussion_r864622745)

@sohhass I believe the reason why the check is removed needs to be better documented. Please create a separate commit. It will also minimize the risk of this line being inadvertently dropped if a conflict occurs when rebasing to a more recent version of LibTomCrypt.

[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/lib/libtomcrypt/src/pk/x25519/x25519_make_key.c:30]: Right, sorry, i didn't look enough in the P-R history.

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by jforissier** [comment]
@sohhass I have merged this PR manually because I wanted to re-order the commits in a sequence that makes more sense in my opinion: the commit which allows to enable `CFG_CRYPTO_X25519=y` needs to come last so that everything is ready at that point. Thanks again for your contribution!

**Comment by sohhass** [comment]
> Thanks again for your contribution!

my pleasure!
